core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -979,6 +1004,21 @@ private void ensureWriteAllowed(Engine.Operation op) throws IllegalIndexShardSta
         }
     }
 
+    private void verifyPrimary() {
+        if (shardRouting.primary() == false) {
+            // must use exception that is not ignore be replication logic. See TransportActions.isShardNotAvailableException
+            throw new UnsupportedOperationException("shard is not a primary " + shardRouting);
+        }
+    }
+
+    private void verifyReplicationTarget() {
+        final IndexShardState state = state();
+        if (shardRouting.primary() && shardRouting.isRelocationTarget() == false && state != IndexShardState.RELOCATED) {
```

Comment:
I guess that's a question of a point of view. To me having "it's not a relocation target" (which we replicate to) is clearer. I'm OK with going your way, if you prefer. Let me know.